File: 650.cpp
/* Project Euler Problem 650: https://projecteuler.net/problem=650 */
// Builds prime-exponent tables and accumulates the divisor-sum sequence modulo 1e9+7.

#include <chrono>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <unordered_map>
#include <utility>
#include <vector>

namespace {

constexpr std::int64_t MOD = 1'000'000'007LL;

// Computes (base^exp) mod MOD using binary exponentiation.
// Runs in O(log exp) time by squaring the base and consuming bits of exp.
// Assumes exp >= 0.
std::int64_t mod_pow(std::int64_t base, std::int64_t exp) {
    std::int64_t result = 1;
    base %= MOD;
    while (exp > 0) {
        if (exp & 1LL) {
            result = (result * base) % MOD;
        }
        base = (base * base) % MOD;
        exp >>= 1LL;
    }
    return result;
}

std::vector<int> build_spf(int limit) {
    std::vector<int> spf(limit + 1);
    for (int i = 0; i <= limit; ++i) {
        spf[i] = i;
    }
    if (limit >= 1) {
        spf[1] = 1;
    }

    int root = static_cast<int>(std::sqrt(limit));
    for (int i = 2; i <= root; ++i) {
        if (spf[i] == i) {
            for (int j = i * i; j <= limit; j += i) {
                if (spf[j] == j) {
                    spf[j] = i;
                }
            }
        }
    }
    return spf;
}

std::vector<std::pair<int, int>> factorize_with_spf(int x, const std::vector<int>& spf) {
    std::vector<std::pair<int, int>> out;
    while (x > 1) {
        int p = spf[x];
        int c = 0;
        while (x % p == 0) {
            x /= p;
            ++c;
        }
        out.push_back({p, c});
    }
    return out;
}

std::int64_t solve(int n) {
    if (n <= 0) {
        return 0;
    }

    std::vector<int> spf = build_spf(n);
    std::vector<int> primes;
    primes.reserve(n / std::max(1.0, std::log(n)));
    for (int i = 2; i <= n; ++i) {
        if (spf[i] == i) {
            primes.push_back(i);
        }
    }

    std::unordered_map<int, int> idx_of;
    idx_of.reserve(primes.size() * 2);
    for (int i = 0; i < static_cast<int>(primes.size()); ++i) {
        idx_of[primes[i]] = i;
    }

    int m = static_cast<int>(primes.size());
    std::vector<std::int64_t> a(m, 0);
    std::vector<std::int64_t> f(m, 0);
    std::vector<std::int64_t> inv_pm1(m, 0);
    for (int i = 0; i < m; ++i) {
        inv_pm1[i] = mod_pow(primes[i] - 1, MOD - 2);
    }

    std::int64_t total = 0;
    int active = 0;

    for (int k = 1; k <= n; ++k) {
        if (k >= 2 && spf[k] == k) {
            ++active;
        }

        for (const auto& [p, v] : factorize_with_spf(k, spf)) {
            a[idx_of[p]] += v;
        }

        std::int64_t d_k = 1;
        std::int64_t kp1 = static_cast<std::int64_t>(k) + 1;

        for (int i = 0; i < active; ++i) {
            std::int64_t ai = a[i];
            std::int64_t fi = f[i] + ai;
            f[i] = fi;
            std::int64_t e = kp1 * ai - 2 * fi;
            if (e != 0) {
                std::int64_t p = primes[i];
                std::int64_t term = (mod_pow(p, e + 1) - 1 + MOD) % MOD;
                term = (term * inv_pm1[i]) % MOD;
                d_k = (d_k * term) % MOD;
            }
        }

        total += d_k;
        if (total >= MOD) {
            total -= MOD;
        }
    }

    return total;
}

}  // namespace

int main() {
    const auto start = std::chrono::steady_clock::now();
    const auto ans = solve(20000);
    const auto end = std::chrono::steady_clock::now();
    const std::chrono::duration<double> elapsed = end - start;

    std::cout << ans << '\n';
    std::cout << "Elapsed: " << elapsed.count() << " s\n";
    return 0;
}

File: 650.py
from __future__ import annotations

"""Project Euler Problem 650: https://projecteuler.net/problem=650

Definitions:
- B(n) = product(C(n, k), k=0..n)
- D(n) = sum of positive divisors of B(n)
- S(N) = sum(D(i), i=1..N) mod MOD

This implementation avoids constructing huge integers and instead works with
prime exponents.
"""

# Builds prime-exponent tables and accumulates the divisor-sum sequence modulo 1e9+7.

MOD = 1_000_000_007


def build_spf(limit: int) -> list[int]:
    """Smallest prime factor table for fast factorizations."""
    spf = list(range(limit + 1))
    if limit >= 1:
        spf[1] = 1
    for i in range(2, int(limit**0.5) + 1):
        if spf[i] == i:
            step = i
            start = i * i
            for j in range(start, limit + 1, step):
                if spf[j] == j:
                    spf[j] = i
    return spf


def factorize_with_spf(x: int, spf: list[int]) -> dict[int, int]:
    """Return prime exponent map for x using precomputed spf."""
    out: dict[int, int] = {}
    while x > 1:
        p = spf[x]
        c = 0
        while x % p == 0:
            x //= p
            c += 1
        out[p] = c
    return out


def solve(n: int) -> int:
    """Return S(n) = sum(D(i), i=1..n) mod MOD."""
    if n <= 0:
        return 0

    spf = build_spf(n)
    primes = [i for i in range(2, n + 1) if spf[i] == i]
    idx_of = {p: i for i, p in enumerate(primes)}
    m = len(primes)

    # a[p] = v_p(k!), where k is current n.
    # f[p] = sum_{i=1..k} v_p(i!).
    a = [0] * m
    f = [0] * m

    # Modular inverses for denominator in sigma(p^e) = (p^(e+1)-1)/(p-1).
    inv_pm1 = [pow(p - 1, MOD - 2, MOD) for p in primes]

    total = 0
    active = 0  # number of primes <= current n

    for k in range(1, n + 1):
        if spf[k] == k and k >= 2:
            active += 1

        for p, v in factorize_with_spf(k, spf).items():
            a[idx_of[p]] += v

        d_k = 1
        kp1 = k + 1
        for i in range(active):
            ai = a[i]
            fi = f[i] + ai
            f[i] = fi
            e = kp1 * ai - 2 * fi
            if e:
                p = primes[i]
                term = (pow(p, e + 1, MOD) - 1) * inv_pm1[i]
                d_k = (d_k * (term % MOD)) % MOD

        total += d_k
        if total >= MOD:
            total -= MOD

    return total


if __name__ == "__main__":

    print(solve(20000))
